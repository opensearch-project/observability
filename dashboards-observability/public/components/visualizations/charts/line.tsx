/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { take, merge } from 'lodash';
import { Plt } from '../plotly/plot';
import { PLOTLY_COLOR } from '../../../../common/constants/shared';

export const Line = ({ visualizations, lineConfig = {}, layoutConfig = {} }: any) => {
  const {
    data,
    metadata: { fields },
  } = visualizations;
  const lineLength = fields.length - 1;
  const lineValues = take(fields, lineLength).map((field: any) => {
    return {
      x: data[fields[lineLength].name],
      y: data[field.name],
      type: 'line',
      name: field.name,
    };
  });

  const config = {
    barmode: 'line',
    xaxis: {
      automargin: true,
    },
    yaxis: {
      automargin: true,
    },
  };
  const lineLayoutConfig = merge(config, layoutConfig);

  return (
    <Plt
      data={lineValues}
      layout={{
        colorway: PLOTLY_COLOR,
        plot_bgcolor: 'rgba(0, 0, 0, 0)',
        paper_bgcolor: 'rgba(0, 0, 0, 0)',
        xaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true,
        },
        yaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true,
        },
        ...lineLayoutConfig,
      }}
      config={lineConfig}
    />
  );
};
