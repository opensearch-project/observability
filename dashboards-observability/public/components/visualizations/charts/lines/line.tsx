/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { find, isEmpty, last, forEach } from 'lodash';
import React, { useMemo } from 'react';
import { AGGREGATIONS, GROUPBY } from '../../../../../common/constants/explorer';
import {
  DEFAULT_CHART_STYLES,
  FILLOPACITY_DIV_FACTOR,
  PLOTLY_COLOR,
  VIS_CHART_TYPES,
  PLOT_MARGIN,
} from '../../../../../common/constants/shared';
import { IVisualizationContainerProps } from '../../../../../common/types/explorer';
import { hexToRgb } from '../../../../components/event_analytics/utils/utils';
import { AvailabilityUnitType } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_availability';
import { ThresholdUnitType } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_thresholds';
import { VisCanvassPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components';
import { Plt } from '../../plotly/plot';
import { transformPreprocessedDataToTraces, preprocessJsonData } from '../shared/common';

export const Line = ({ visualizations, layout, config }: any) => {
  const {
    DefaultModeLine,
    Interpolation,
    LineWidth,
    FillOpacity,
    MarkerSize,
    LegendPosition,
    ShowLegend,
    DefaultModeScatter,
    LabelAngle,
  } = DEFAULT_CHART_STYLES;

  const {
    data: {
      rawVizData: {
        data: queriedVizData,
        jsonData,
        metadata: { fields },
      },
      userConfigs: {
        dataConfig: {
          chartStyles = {},
          legend = {},
          span = {},
          colorTheme = [],
          thresholds = [],
          tooltipOptions = {},
          panelOptions = {},
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
          breakdowns = [],
        } = {},
        layoutConfig = {},
        availabilityConfig = {},
      } = {},
    },
    vis: { icontype, name },
  }: IVisualizationContainerProps = visualizations;

  const tooltipMode =
    tooltipOptions.tooltipMode !== undefined ? tooltipOptions.tooltipMode : 'show';
  const tooltipText = tooltipOptions.tooltipText !== undefined ? tooltipOptions.tooltipText : 'all';
  const lastIndex = fields.length - 1;
  const visType: string = name;
  const mode =
    chartStyles.style || (visType === VIS_CHART_TYPES.Line ? DefaultModeLine : DefaultModeScatter);
  const lineShape = chartStyles.interpolation || Interpolation;
  const lineWidth = chartStyles.lineWidth || LineWidth;
  const showLegend = !(legend.showLegend && legend.showLegend !== ShowLegend);
  const legendPosition = legend.position || LegendPosition;
  const markerSize = chartStyles.pointSize || MarkerSize;
  const fillOpacity =
    chartStyles.fillOpacity !== undefined
      ? chartStyles.fillOpacity / FILLOPACITY_DIV_FACTOR
      : FillOpacity / FILLOPACITY_DIV_FACTOR;
  const tickAngle = chartStyles.rotateLabels || LabelAngle;
  const labelSize = chartStyles.labelSize;
  const legendSize = legend.legendSize;
  const getSelectedColorTheme = (field: any, index: number) =>
    (colorTheme.length > 0 &&
      colorTheme.find((colorSelected) => colorSelected.name.name === field)?.color) ||
    PLOTLY_COLOR[index % PLOTLY_COLOR.length];
  const timestampField = find(fields, (field) => field.type === 'timestamp');
  let xaxis = [timestampField];

  if (isEmpty(series)) return <VisCanvassPlaceholder message='Missing mandatory series' icon={icontype} />;

  const addStylesToTraces = (traces, traceStyles) => {
    const { fillOpacity, tooltipMode, tooltipText, lineWidth, lineShape, markerSize } = traceStyles;
    return traces.map((trace, idx: number) => {
      const selectedColor = getSelectedColorTheme(trace.aggName, idx);
      const fillColor = hexToRgb(selectedColor, fillOpacity);

      return {
        ...trace,
        hoverinfo: tooltipMode === 'hidden' ? 'none' : tooltipText,
        type: 'line',
        mode,
        ...{
          fill: 'tozeroy',
          fillcolor: fillColor,
        },
        line: {
          shape: lineShape,
          width: lineWidth,
          color: selectedColor,
        },
        marker: {
          size: markerSize,
          ...{
            color: fillColor,
            line: {
              color: selectedColor,
              width: lineWidth,
            },
          },
        },
      };
    });
  };

  let lines = useMemo(() => {
    const visConfig = {
      dimensions,
      series,
      breakdowns,
      span,
    };
    const traceStyles = {
      fillOpacity,
      tooltipMode,
      tooltipText,
      lineShape,
      lineWidth,
      markerSize,
    };

    return addStylesToTraces(
      transformPreprocessedDataToTraces(preprocessJsonData(jsonData, visConfig), visConfig),
      traceStyles
    );
  }, [
    chartStyles,
    jsonData,
    dimensions,
    series,
    span,
    breakdowns,
    panelOptions,
    tooltipOptions,
  ]);

  const mergedLayout = useMemo(() => {
    const axisLabelsStyle = {
      automargin: true,
      tickfont: {
        ...(labelSize && {
          size: labelSize,
        }),
      },
    };

    return {
      ...layout,
      title: panelOptions.title || layoutConfig.layout?.title || '',
      legend: {
        ...layout.legend,
        orientation: legendPosition,
        ...(legendSize && {
          font: {
            size: legendSize,
          },
        }),
      },
      autosize: true,
      xaxis: {
        tickangle: tickAngle,
        ...axisLabelsStyle,
      },
      yaxis: {
        ...axisLabelsStyle,
      },
      showlegend: showLegend,
      margin: PLOT_MARGIN,
    };
  }, [visualizations]);

  if (thresholds || availabilityConfig.level) {
    const thresholdTraces = {
      x: [],
      y: [],
      mode: 'text',
      text: [],
      showlegend: false,
    };

    const levels = availabilityConfig.level ? availabilityConfig.level : [];

    const mapToLine = (list: ThresholdUnitType[] | AvailabilityUnitType[], lineStyle: any) => {
      return list.map((thr: ThresholdUnitType) => {
        thresholdTraces.x.push(
          queriedVizData[
            !isEmpty(xaxis) ? xaxis[xaxis.length - 1]?.name : fields[lastIndex].name
          ][0]
        );
        thresholdTraces.y.push(thr.value * (1 + 0.06));
        thresholdTraces.text.push(thr.name);
        return {
          type: 'line',
          x0: queriedVizData[!isEmpty(xaxis) ? xaxis[0]?.name : fields[lastIndex].name][0],
          y0: thr.value,
          x1: last(queriedVizData[!isEmpty(xaxis) ? xaxis[0]?.name : fields[lastIndex].name]),
          y1: thr.value,
          name: thr.name || '',
          opacity: 0.7,
          line: {
            color: thr.color,
            width: 3,
            ...lineStyle,
          },
        };
      });
    };

    mergedLayout.shapes = [
      ...mapToLine(
        thresholds.filter((i: ThresholdUnitType) => i.value !== ''),
        { dash: 'dashdot' }
      ),
      ...mapToLine(levels, {}),
    ];

    lines = [...lines, thresholdTraces];
  }

  const mergedConfigs = useMemo(
    () => ({
      ...config,
      ...(layoutConfig.config && layoutConfig.config),
    }),
    [config, layoutConfig.config]
  );

  return <Plt data={lines} layout={mergedLayout} config={mergedConfigs} />;
};
