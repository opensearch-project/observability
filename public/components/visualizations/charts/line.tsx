/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Plt } from '../plotly/plot';

export const Line = ({
  xvalues,
  yvalues,
  name,
  layoutConfig,
}: any) => {
  
  return (
    <Plt 
      data={[
        {
          x: xvalues,
          y: yvalues,
          type: 'lines',
          name,
        }
      ]}
      layout={{
        plot_bgcolor: 'rgba(0, 0, 0, 0)',
        paper_bgcolor: 'rgba(0, 0, 0, 0)',
        xaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true,
        },
        yaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true
        },
        ...layoutConfig
      }}
    />  
  );
};