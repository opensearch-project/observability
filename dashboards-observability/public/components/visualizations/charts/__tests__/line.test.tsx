/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { Line } from '../line';
import { 
  LAYOUT_CONFIG,
  SAMPLE_VISUALIZATIONS
} from '../../../../../test/event_analytics_constants';

describe('Line component', () => {
  configure({ adapter: new Adapter() });

  it('Renders line component', async () => {
    
    const wrapper = mount(
      <Line 
        visualizations={SAMPLE_VISUALIZATIONS}
        name="Event counts"
        layoutConfig={LAYOUT_CONFIG}
        isUniColor={true}
      />
    );
    
    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});