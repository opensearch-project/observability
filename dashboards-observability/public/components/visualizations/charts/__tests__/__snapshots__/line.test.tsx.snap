// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line component Renders line component 1`] = `
<Line
  isUniColor={true}
  layoutConfig={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  name="Event counts"
  visualizations={
    Object {
      "data": Object {
        "count()": Array [
          2549,
          9337,
          1173,
        ],
        "span(timestamp,1M)": Array [
          "2021-05-01 00:00:00",
          "2021-06-01 00:00:00",
          "2021-07-01 00:00:00",
        ],
      },
      "jsonData": Array [
        Object {
          "count()": 2549,
          "span(timestamp,1M)": "2021-05-01 00:00:00",
        },
        Object {
          "count()": 9337,
          "span(timestamp,1M)": "2021-06-01 00:00:00",
        },
        Object {
          "count()": 2549,
          "span(timestamp,1M)": "2021-07-01 00:00:00",
        },
      ],
      "metadata": Object {
        "fields": Array [
          Object {
            "name": "count()",
            "type": "integer",
          },
          Object {
            "name": "span(timestamp,1M)",
            "type": "timestamp",
          },
        ],
      },
    }
  }
>
  <Plt
    config={Object {}}
    data={
      Array [
        Object {
          "name": "count()",
          "type": "line",
          "x": Array [
            "2021-05-01 00:00:00",
            "2021-06-01 00:00:00",
            "2021-07-01 00:00:00",
          ],
          "y": Array [
            2549,
            9337,
            1173,
          ],
        },
      ]
    }
    layout={
      Object {
        "barmode": "line",
        "colorway": Array [
          "#8C55A3",
        ],
        "height": 220,
        "margin": Object {
          "b": 15,
          "l": 60,
          "pad": 0,
          "r": 10,
          "t": 30,
        },
        "paper_bgcolor": "rgba(0, 0, 0, 0)",
        "plot_bgcolor": "rgba(0, 0, 0, 0)",
        "showlegend": true,
        "xaxis": Object {
          "automargin": true,
        },
        "yaxis": Object {
          "automargin": true,
        },
      }
    }
  >
    <PlotlyComponent
      config={
        Object {
          "displayModeBar": false,
        }
      }
      data={
        Array [
          Object {
            "name": "count()",
            "type": "line",
            "x": Array [
              "2021-05-01 00:00:00",
              "2021-06-01 00:00:00",
              "2021-07-01 00:00:00",
            ],
            "y": Array [
              2549,
              9337,
              1173,
            ],
          },
        ]
      }
      debug={false}
      layout={
        Object {
          "autosize": true,
          "barmode": "line",
          "colorway": Array [
            "#8C55A3",
          ],
          "height": 220,
          "hovermode": "closest",
          "legend": Object {
            "orientation": "h",
            "traceorder": "normal",
          },
          "margin": Object {
            "b": 15,
            "l": 60,
            "pad": 0,
            "r": 10,
            "t": 30,
          },
          "paper_bgcolor": "rgba(0, 0, 0, 0)",
          "plot_bgcolor": "rgba(0, 0, 0, 0)",
          "showlegend": true,
          "xaxis": Object {
            "automargin": true,
          },
          "yaxis": Object {
            "automargin": true,
          },
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useResizeHandler={true}
    >
      <div
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </PlotlyComponent>
  </Plt>
</Line>
`;

exports[`Line component Renders line component 2`] = `
<Line
  isUniColor={true}
  layoutConfig={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  name="Event counts"
  visualizations={
    Object {
      "data": Object {
        "count()": Array [
          2549,
          9337,
          1173,
        ],
        "span(timestamp,1M)": Array [
          "2021-05-01 00:00:00",
          "2021-06-01 00:00:00",
          "2021-07-01 00:00:00",
        ],
      },
      "jsonData": Array [
        Object {
          "count()": 2549,
          "span(timestamp,1M)": "2021-05-01 00:00:00",
        },
        Object {
          "count()": 9337,
          "span(timestamp,1M)": "2021-06-01 00:00:00",
        },
        Object {
          "count()": 2549,
          "span(timestamp,1M)": "2021-07-01 00:00:00",
        },
      ],
      "metadata": Object {
        "fields": Array [
          Object {
            "name": "count()",
            "type": "integer",
          },
          Object {
            "name": "span(timestamp,1M)",
            "type": "timestamp",
          },
        ],
      },
    }
  }
>
  <Plt
    config={Object {}}
    data={
      Array [
        Object {
          "name": "count()",
          "type": "line",
          "x": Array [
            "2021-05-01 00:00:00",
            "2021-06-01 00:00:00",
            "2021-07-01 00:00:00",
          ],
          "y": Array [
            2549,
            9337,
            1173,
          ],
        },
      ]
    }
    layout={
      Object {
        "barmode": "line",
        "colorway": Array [
          "#8C55A3",
        ],
        "height": 220,
        "margin": Object {
          "b": 15,
          "l": 60,
          "pad": 0,
          "r": 10,
          "t": 30,
        },
        "paper_bgcolor": "rgba(0, 0, 0, 0)",
        "plot_bgcolor": "rgba(0, 0, 0, 0)",
        "showlegend": true,
        "xaxis": Object {
          "automargin": true,
          "autorange": true,
          "range": Array [
            "2021-04-27 03:45:50.6702",
            "2021-07-04 20:14:09.3298",
          ],
          "type": "date",
        },
        "yaxis": Object {
          "automargin": true,
          "autorange": true,
          "range": Array [
            425.69084628670123,
            10084.309153713299,
          ],
          "type": "linear",
        },
      }
    }
  >
    <PlotlyComponent
      config={
        Object {
          "displayModeBar": false,
        }
      }
      data={
        Array [
          Object {
            "name": "count()",
            "type": "line",
            "x": Array [
              "2021-05-01 00:00:00",
              "2021-06-01 00:00:00",
              "2021-07-01 00:00:00",
            ],
            "y": Array [
              2549,
              9337,
              1173,
            ],
          },
        ]
      }
      debug={false}
      layout={
        Object {
          "autosize": true,
          "barmode": "line",
          "colorway": Array [
            "#8C55A3",
          ],
          "height": 220,
          "hovermode": "closest",
          "legend": Object {
            "orientation": "h",
            "traceorder": "normal",
          },
          "margin": Object {
            "b": 15,
            "l": 60,
            "pad": 0,
            "r": 10,
            "t": 30,
          },
          "paper_bgcolor": "rgba(0, 0, 0, 0)",
          "plot_bgcolor": "rgba(0, 0, 0, 0)",
          "showlegend": true,
          "xaxis": Object {
            "automargin": true,
            "autorange": true,
            "range": Array [
              "2021-04-27 03:45:50.6702",
              "2021-07-04 20:14:09.3298",
            ],
            "type": "date",
          },
          "yaxis": Object {
            "automargin": true,
            "autorange": true,
            "range": Array [
              425.69084628670123,
              10084.309153713299,
            ],
            "type": "linear",
          },
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useResizeHandler={true}
    >
      <div
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </PlotlyComponent>
  </Plt>
</Line>
`;
